pkg/controllers/v1alpha1/fluidapp/dataflowaffinity/dataflowaffinity_controller.go

```diff
@@ -0,0 +1,201 @@
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/record"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	"github.com/fluid-cloudnative/fluid/pkg/ctrl/watch"
+	"github.com/fluid-cloudnative/fluid/pkg/utils/kubeclient"
+)
+
+const DataOpJobControllerName string = "DataOpJobController"
+
+type DataOpJobReconciler struct {
+	client.Client
+	Recorder record.EventRecorder
+	Log      logr.Logger
+}
+
+func (f *DataOpJobReconciler) ControllerName() string {
+	return DataOpJobControllerName
+}
+
+func (f *DataOpJobReconciler) ManagedResource() client.Object {
+	return &batchv1.Job{}
+}
+
+type reconcileRequestContext struct {
+	context.Context
+	Log logr.Logger
+	job *batchv1.Job
+	types.NamespacedName
+}
+
+func NewDataOpJobReconciler(client client.Client,
+	log logr.Logger,
+	recorder record.EventRecorder) *DataOpJobReconciler {
+	return &DataOpJobReconciler{
+		Client:   client,
+		Recorder: recorder,
+		Log:      log,
+	}
+}
+
+// Reconcile reconciles Jobs
+// +kubebuilder:rbac:groups=v1,resources=pods,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=v1,resources=pods/status,verbs=get;update;patch
+func (f *DataOpJobReconciler) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	requestCtx := reconcileRequestContext{
+		Context:        ctx,
+		Log:            f.Log.WithValues("namespacedName", request.NamespacedName),
+		NamespacedName: request.NamespacedName,
+	}
+	job, err := kubeclient.GetJob(f.Client, request.Name, request.Namespace)
+	if err != nil {
+		requestCtx.Log.Error(err, "fetch job error")
+		return utils.RequeueIfError(err)
+	}
+	if job == nil {
+		requestCtx.Log.Info("job not found", "name", request.Name, "namespace", request.Namespace)
+		return utils.NoRequeue()
+	}
+	requestCtx.job = job
+
+	if !watch.JobShouldInQueue(job) {
+		requestCtx.Log.Info("job should not in queue", "name", request.Name, "namespace", request.Namespace)
+		return utils.NoRequeue()
+	}
+	if job.Annotations == nil {
+		job.Annotations = map[string]string{}
+	}
+
+	// inject dataflow enabled affinity if not exist.
+	if _, ok := job.Annotations[common.AnnotationDataFlowAffinityInject]; !ok {
+		job.Annotations[common.AnnotationDataFlowAffinityInject] = "true"
+		if err := f.Client.Update(ctx, job); err != nil {
+			requestCtx.Log.Error(err, "Failed to add dataflow affinity enabled label", "AnnotationUpdateError", ctx)
+			return utils.RequeueIfError(err)
+		}
+	}
+	// get job' status, if succeed, add label to job.
+	condition := kubeclient.GetFinishedJobCondition(job)
+	if condition != nil && condition.Type == batchv1.JobComplete {
+		err = f.injectPodNodeLabelsToJob(job)
+		if err != nil {
+			requestCtx.Log.Error(err, "update labels for job failed")
+			return utils.RequeueIfError(err)
+		}
+	}
```

Comment:
How about using `f.Client.Patch()` to patch annotations to the jobs? Current code calls two consecutive `f.Client.Update()` (one in line 102 and the other inside `f.injectPodNodeLabelsToJob`), so in my opinion the latter call will definitely get a conflict error. I think we better avoid this.

Reply:
Updating label (`f.injectPodNodeLabelsToJob`) occurs only when job succeed(line 109), so it is not possible to call update twice in once reconciler.